core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -132,15 +132,20 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
       AppStatusStore.CURRENT_VERSION, logDir.toString())
 
     try {
-      open(new File(path, "listing.ldb"), metadata)
+      open(dbPath, metadata)
     } catch {
       // If there's an error, remove the listing database and any existing UI database
       // from the store directory, since it's extremely likely that they'll all contain
       // incompatible information.
       case _: UnsupportedStoreVersionException | _: MetadataMismatchException =>
         logInfo("Detected incompatible DB versions, deleting...")
         path.listFiles().foreach(Utils.deleteRecursively)
-        open(new File(path, "listing.ldb"), metadata)
+        open(dbPath, metadata)
+      case e: Exception =>
+        // Get rid of the old data and re-create it. The store is either old or corrupted.
+        logWarning(s"Failed to load disk store $path for app listing.", e)
```

Comment:
Is there a danger here that a transient error causes this to delete an otherwise valid file?

Reply:
No, the file listing.ldb is built from event logs, which stores list of applications(history server home page). It is used like cache.

Reply:
Shall we merge it into the previous case?

Reply:
@cloud-fan No. 
Currently history server writes to a leveldb file for each application(if its page is accessed). And there will be one db called listing.ldb which stores the listing.
If the version is not matched, then all the db files should be deleted. 
In this PR, if only the listing db is corrupted, then just delete the listing db and recreate it.

Reply:
I see, shall we catch `DBException` instead of the general `Exception`?

Reply:
Yes I thought about catch `DBException`
But in this pr I just follows opening disk store for app:https://github.com/apache/spark/blob/0ffa7c488fa8156e2a1aa282e60b7c36b86d8af8/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala#L307